### ISSUE TITLE
automaton corruption during `knuthbendix!`?

### ISSUE THREAD

**Issue by kalmarek** [kalmarek/KnuthBendix.jl#46]
on `master`:
```julia
function RWS_Example_237_abaB(n)
    Al = Alphabet(['a', 'b', 'B'])
    KnuthBendix.set_inversion!(Al, 'b', 'B')

    a, b, B = Word.([i] for i in 1:3)
    ε = one(a)

    eqns = [b*B=>ε, B*b=>ε, a^2=>ε, b^3=>ε, (a*b)^7=>ε, (a*b*a*B)^n=>ε]

    R = RewritingSystem(eqns, LenLex(Al), bare=true)
    return R
end
```
this example collapses when using automata (but it shouldn't):
```julia
julia> R = RWS_Example_237_abaB(4)
Rewriting System with 6 rules ordered by LenLex{Char}(Alphabet{Char}['a', 'b', 'B']):
   1 ✓ b*B   →  (empty word)
   2 ✓ B*b   →  (empty word)
   3 ✓ a*a   →  (empty word)
   4 ✓ b*b*b     →  (empty word)
   5 ✓ a*b*a*b*a*b*a*b*a*b*a*b*a*b   →  (empty word)
   6 ✓ a*b*a*B*a*b*a*B*a*b*a*B*a*b*a*B   →  (empty word)

julia> aut = KnuthBendix.knuthbendix(R, maxrules=35, implementation=:automata)
Rewriting System with 3 rules ordered by LenLex{Char}(Alphabet{Char}['a', 'b', 'B']):
   1 ✓ a     →  (empty word)
   2 ✓ b     →  (empty word)
   3 ✓ B     →  (empty word)

julia> kbs2 = KnuthBendix.knuthbendix(R, maxrules=35, implementation=:naive_kbs2)
┌ Warning: Maximum number of rules (35) in the RewritingSystem reached.
│                 You may retry with `maxrules` kwarg set to higher value.
└ @ KnuthBendix ~/.julia/dev/KnuthBendix/src/kbs.jl:67
Rewriting System with 36 rules ordered by LenLex{Char}(Alphabet{Char}['a', 'b', 'B']):
   1 ✓ a*a   →  (empty word)
   2 ✓ B*b   →  (empty word)
   3 ✓ b*B   →  (empty word)
   4 ✓ b*b   →  B
   5 ✓ B*B   →  b
   6 ✓ b*a*b*a*b*a*b     →  a*B*a*B*a*B*a
   7 ✓ b*a*b*a*B*a*b*a*B     →  B*a*B*a*b*a*B*a
   8 ✓ b*a*B*a*b*a*B*a   →  a*b*a*B*a*b*a*B
   9 ✓ a*B*a*b*a*b*a*B*a*B*a*b   →  B*a*b*a*B*a*B*a*b*a*B
  10 ✓ B*a*B*a*b*a*b*a*B*a*B*a   →  b*a*b*a*B*a*B*a*b*a*b
  11 ✓ b*a*b*a*B*a*b*a*b*a*B*a   →  B*a*B*a*B*a*b*a*b*a*b
  12 ✓ b*a*b*a*B*a*B*a*b*a*b*a   →  B*a*B*a*b*a*b*a*B*a*B
  13 ✓ b*a*B*a*b*a*b*a*B*a*b*a   →  B*a*b*a*b*a*B*a*B*a*b
  14 ✓ a*B*a*B*a*B*a*b   →  b*a*b*a*b*a*B
  15 ✓ b*a*B*a*B*a*B*a   →  B*a*b*a*b*a*b
  16 ✓ b*a*b*a*b*a*B*a*B*a*b     →  a*B*a*b*a*b*a*B*a*b*a
  17 ✓ B*a*B*a*b*a*B*a*b     →  b*a*b*a*B*a*b*a
  18 ✓ B*a*B*a*b*a*B*a*B     →  b*a*b*a*B*a*b*a*b
  19 ✓ B*a*b*a*B*a*b*a   →  a*B*a*b*a*B*a*b
  20 ✓ B*a*B*a*b*a*b*a*B*a*b     →  a*b*a*b*a*B*a*B*a*b*a
  21 ✓ b*a*b*a*B*a*B*a*b*a*B     →  a*B*a*B*a*b*a*b*a*B*a
  22 ✓ a*B*a*b*a*B*a*B*a*b*a*B   →  B*a*b*a*b*a*B*a*B*a*b
  23 ✓ B*a*b*a*B*a*B*a*b*a*b     →  a*B*a*b*a*b*a*B*a*B*a
  24 ✓ a*b*a*b*a*B*a*B*a*b*a*b   →  B*a*B*a*b*a*b*a*B*a*B
  25 ✓ b*a*B*a*B*a*b*a*b*a*b     →  a*b*a*B*a*b*a*b*a*B*a
  26 ✓ B*a*b*a*b*a*B*a*B*a*B     →  a*B*a*b*a*B*a*B*a*b*a
  27 ✓ b*a*B*a*b*a*b*a*B*a*B     →  a*b*a*B*a*B*a*b*a*b*a
  28 ✓ B*a*b*a*b*a*b*a   →  b*a*B*a*B*a*B
  29 ✓ B*a*b*a*b*a*B*a*B*a*b*a   →  b*a*B*a*b*a*b*a*B*a*b
  30 ✓ B*a*B*a*B*a*b*a*b*a*B*a   →  a*b*a*B*a*B*a*b*a*B
  31 ✓ B*a*B*a*B*a*B     →  a*b*a*b*a*b*a
  32 ✓ a*b*a*b*a*b*a*B   →  B*a*B*a*B*a*b
  33 ✓ b*a*b*a*b*a*B*a*b*a*b*a*b     →  a*b*a*b*a*B*a*b*a*b*a
  34 ✓ a*b*a*B*a*b*a*b*a*B*a*b   →  b*a*B*a*B*a*b*a*b*a*B
  35 ✓ a*B*a*b*a*b*a*B*a*b*a*b   →  b*a*b*a*b*a*B*a*B*a*B
  36 ✓ b*a*B*a*B*a*b*a*b*a*B*a*B     →  a*B*a*b*a*b*a*B*a*B*a
```

however stopping the process in the middle makes `knuthbendix!` with automaton complete successfully: 

```julia
julia> let n = 34
           kbs = KnuthBendix.knuthbendix(R, maxrules=n, implementation=:naive_kbs2)
           aut = KnuthBendix.knuthbendix(R, maxrules=n, implementation=:automata)
           @info "" length(kbs) length(aut)

           @info "" length(KnuthBendix.knuthbendix(aut, implementation=:automata)) length(KnuthBendix.knuthbendix(aut, implementation=:naive_kbs2))
       end
┌ Warning: Maximum number of rules (34) in the RewritingSystem reached.
│                 You may retry with `maxrules` kwarg set to higher value.
└ @ KnuthBendix ~/.julia/dev/KnuthBendix/src/kbs.jl:67
┌ Warning: Maximum number of rules (34) in the RewritingSystem reached.
│                 You may retry with `maxrules` kwarg set to higher value.
└ @ KnuthBendix ~/.julia/dev/KnuthBendix/src/kbs.jl:166
┌ Info: 
│   length(kbs) = 35
└   length(aut) = 35
┌ Info: 
│   length(KnuthBendix.knuthbendix(aut, implementation = :automata)) = 40
└   length(KnuthBendix.knuthbendix(aut, implementation = :naive_kbs2)) = 40
```

**Comment by mikolajpabiszczak** [comment]
Ahh, another nightmare to add to this week. Btw: have you tested other implementations (e.g. the one with deleting)? Are they ok and we know that this is only for the automata implmementation?

**Comment by kalmarek** [comment]
yes, I did ;)

I actually stumbled upon this bug after a change in `RewritingSystem` (adding relations for inverses in its alphabet). This changed the order of initial relations and triggered the bug:
```julia
eqns = [B*b=>ε, b*B=>ε, a^2=>ε, b^3=>ε, (a*b)^7=>ε, (a*b*a*B)^n=>ε] # old order, :automata works

eqns = [b*B=>ε, B*b=>ε, a^2=>ε, b^3=>ε, (a*b)^7=>ε, (a*b*a*B)^n=>ε] # new order, :automata fails
```
the size of the confluent system should be `40`.
what is really interesting it doesn't matter what is `n` (the stopping point) in the second snippet. A truly bizzare bug.

**Comment by mikolajpabiszczak** [comment]
This is kind of a nightmare. I will try to take a look at that in the weekend.

**Comment by mikolajpabiszczak** [comment]
Ok, unfortunately I was not able to find the reason, although something is telling me that the problem lies in the `makeindexautomaton` function...

**Comment by kalmarek** [comment]
fixed by #54
